crypto/params.c

```diff
@@ -0,0 +1,191 @@
+    return 1;
+}
+
+static int OSSL_PARAM_set_int_common(const OSSL_PARAM *p, const char *key,
+                                     const void *val, unsigned long type)
+{
+    size_t width;
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL)
+        return 0;
+
+    /* Type safety. */
+    width = OSSL_PARAM_get_width(type);
+    if (p->data_type != type || p->buffer_size != width) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_SET_INT_COMMON, CRYPTO_R_TYPE_MISMATCH);
+        return 0;
+    }
+
+    memcpy(p->buffer, val, width);
+    if (p->return_size != NULL)
+        *p->return_size = width;
+    return 1;
+}
+
+#define PARAM_INT(name, type, PARAM) \
+    int OSSL_PARAM_get_##name(const OSSL_PARAM *p, const char *key, type *val) \
+    { \
+        return OSSL_PARAM_get_int_common(p, key, val, PARAM); \
+    } \
+    int OSSL_PARAM_set_##name(const OSSL_PARAM *p, const char *key, type val) \
+    { \
+        return OSSL_PARAM_set_int_common(p, key, &val, PARAM); \
+    }
+
+PARAM_INT(int, int, OSSL_PARAM_INT)
+PARAM_INT(uint, unsigned int, OSSL_PARAM_UINT)
+PARAM_INT(int64, int64_t, OSSL_PARAM_INT64)
+PARAM_INT(uint64, uint64_t, OSSL_PARAM_UINT64)
+PARAM_INT(long, long int, OSSL_PARAM_LONG)
+PARAM_INT(ulong, unsigned long int, OSSL_PARAM_ULONG)
+PARAM_INT(size_t, size_t, OSSL_PARAM_SIZET)
+
+/*
+ * In the get/set double functions, "width" is always sizeof(double) but
+ * having parallel construction with the integer functions above seems
+ * worthwhile.
+ */
+
+int OSSL_PARAM_get_double(const OSSL_PARAM *p, const char *key, double *val)
+{
+    size_t width;
+    double *dp;
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL)
+        return 0;
+
+    /* Type safety. */
+    width = OSSL_PARAM_get_width(OSSL_PARAM_DOUBLE);
+    if (p->data_type != OSSL_PARAM_DOUBLE || p->buffer_size != width) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_GET_DOUBLE, CRYPTO_R_TYPE_MISMATCH);
+        return 0;
+    }
+
+    dp = (double *)p->buffer;
+    *val = *dp;
+    return 1;
+}
+
+int OSSL_PARAM_set_double(const OSSL_PARAM *p, const char *key, double val)
+{
+    size_t width;
+    double *dp;
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL)
+        return 0;
+
+    /* Type safety. */
+    width = OSSL_PARAM_get_width(OSSL_PARAM_DOUBLE);
+    if (p->data_type != OSSL_PARAM_DOUBLE || p->buffer_size != width) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_SET_DOUBLE, CRYPTO_R_TYPE_MISMATCH);
+        return 0;
+    }
+    dp = (double *)p->buffer;
+    *dp = val;
+    if (p->return_size != NULL)
+        *p->return_size = width;
+    return 1;
+}
+
+int OSSL_PARAM_get_BN(const OSSL_PARAM *p, const char *key, BIGNUM **val)
+{
+    BIGNUM *b;
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL)
+        return 0;
+
+    /* Type safety. */
+    if (p->data_type != OSSL_PARAM_BIGNUM) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_GET_BN, CRYPTO_R_TYPE_MISMATCH);
+        return 0;
+    }
+
+    if ((b = BN_native2bn(p->buffer, (int)p->bn_size, *val)) == NULL)
```

Comment:
Depending on who's calling, either `p->buffer_size` or `*p->return_size` will have the right size.  I suggest the same solution I suggested to @paulidale:


``` C
    if ((b = BN_native2bn(p->buffer,
                          p->return_size != NULL && &p->return_size == 0 
                          ? (int)*p->return_size : (int)p->buffer_size,
                          *val)) == NULL)
```
        if (b != NULL) {
            *val = b;
            return 1;
        }
```